File: fortran/src/newton.f90
module newton

implicit none

type, abstract :: newtoncallback
contains
    procedure(dpfunc), deferred :: func
    procedure(dpfunc), deferred :: deriv
end type newtoncallback

abstract interface
    function dpfunc(this, x) result(fx)
        import :: newtoncallback
        class(newtoncallback), intent(in) :: this
        double precision, intent(in) :: x
        double precision :: fx
    end function dpfunc
end interface

contains

double precision function newtonsolver(cb, x0, maxiter, tol)
    class(newtoncallback), intent(in) :: cb
    double precision, intent(in) :: x0
    integer, intent(in), optional :: maxiter
    double precision, intent(in), optional :: tol

    integer :: maxiter_
    double precision :: tol_

    integer :: i
    double precision :: p
    double precision :: p0

    p0 = x0
    maxiter_ = 50
    if (present(maxiter)) maxiter_ = maxiter
    tol_ = 1d-8
    if (present(tol)) tol_ = tol

    do i = 1, maxiter_
        p = p0 - cb.func(p0) / cb.deriv(p0)
        if (abs(p - p0) < tol_) then
            newtonsolver = p
            return
        end if
        p0 = p
    end do

    write(*,*) "Not converged."
    stop 1
end function newtonsolver

end module newton

File: matlab/newton.m
function p = newton(x0, func, deriv, varargin)

switch nargin
    case 3
        maxiter = 50;
        tol = 1e-8;
    case 4
        maxiter = varargin{1};
        tol = 1e-8;
    case 5
        maxiter = varargin{1};
        tol = varargin{2};
end

p0 = x0;
for ii = 1:maxiter
    p = p0 - func(p0)/deriv(p0);
    if abs(p - p0) < tol
        return
    end
    p0 = p;
end
error('Not converged.');
